File: aws-globalaccelerator-accelerator/src/main/java/software/amazon/globalaccelerator/accelerator/CreateHandler.java
package software.amazon.globalaccelerator.accelerator;

import com.amazonaws.services.globalaccelerator.AWSGlobalAccelerator;
import com.amazonaws.services.globalaccelerator.model.Accelerator;
import com.amazonaws.services.globalaccelerator.model.AcceleratorStatus;
import com.amazonaws.services.globalaccelerator.model.CreateAcceleratorRequest;
import com.amazonaws.services.globalaccelerator.model.DescribeAcceleratorRequest;
import com.amazonaws.services.globalaccelerator.model.AcceleratorNotFoundException;
import com.amazonaws.services.globalaccelerator.model.Tag;
import lombok.val;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.ProgressEvent;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class CreateHandler extends BaseHandler<CallbackContext> {
    @Override
    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
        final AmazonWebServicesClientProxy proxy,
        final ResourceHandlerRequest<ResourceModel> request,
        final CallbackContext callbackContext,
        final Logger logger
    ) {
        val agaClient = AcceleratorClientBuilder.getClient();
        val inferredCallbackContext = callbackContext != null ?
                callbackContext :
                CallbackContext.builder()
                        .stabilizationRetriesRemaining(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
                        .build();

        val model = request.getDesiredResourceState();
        if (model.getAcceleratorArn() == null) {
            return createAcceleratorStep(model, request, proxy, agaClient, logger);
        } else {
            return HandlerCommons.waitForSynchronizedStep(inferredCallbackContext, model, proxy, agaClient, logger);
        }
    }

    /**
     * Create an accelerator and create the correct progress continuation context
     */
    private ProgressEvent<ResourceModel, CallbackContext> createAcceleratorStep(final ResourceModel model,
                                                                                final ResourceHandlerRequest<ResourceModel> handlerRequest,
                                                                                final AmazonWebServicesClientProxy proxy,
                                                                                final AWSGlobalAccelerator agaClient,
                                                                                final Logger logger) {
        logger.log("Creating new accelerator.");
        val acc = createAccelerator(model, handlerRequest, proxy, agaClient);
        model.setAcceleratorArn(acc.getAcceleratorArn());


        List<String> allAddresses = null;
        val ipSets = acc.getIpSets();
        if (ipSets != null) {
            allAddresses = ipSets.stream()
                    .flatMap(x -> x.getIpAddresses().stream())
                    .collect(Collectors.toList());
        }
        model.setIpAddresses(allAddresses);

        val callbackContext = CallbackContext.builder()
                .stabilizationRetriesRemaining(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
                .build();

        return ProgressEvent.defaultInProgressHandler(callbackContext, 0, model);
    }

    /**
     * Create the accelerator based on the provided ResourceModel
     */
    private Accelerator createAccelerator(final ResourceModel model,
                                          final ResourceHandlerRequest<ResourceModel> handlerRequest,
                                          final AmazonWebServicesClientProxy proxy,
                                          final AWSGlobalAccelerator agaClient) {
        List<Tag> convertedTags = null;
        if (model.getTags() != null) {
            convertedTags = model.getTags().stream()
                    .map(x -> new Tag().withKey(x.getKey()).withValue(x.getValue()))
                    .collect(Collectors.toList());
        }

        val request = new CreateAcceleratorRequest()
                        .withName(model.getName())
                        .withEnabled(model.getEnabled())
                        .withIpAddresses(model.getIpAddresses())
                        .withTags(convertedTags)
                        .withIdempotencyToken(handlerRequest.getLogicalResourceIdentifier());

        return proxy.injectCredentialsAndInvoke(request, agaClient::createAccelerator).getAccelerator();
    }
}

File: aws-globalaccelerator-accelerator/src/main/kotlin/software/amazon/globalaccelerator/accelerator/CreateHandler.kt
package software.amazon.globalaccelerator.accelerator

import com.amazonaws.services.globalaccelerator.AWSGlobalAccelerator
import com.amazonaws.services.globalaccelerator.model.Accelerator
import com.amazonaws.services.globalaccelerator.model.CreateAcceleratorRequest
import com.amazonaws.services.globalaccelerator.model.Tag
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.ProgressEvent
import software.amazon.cloudformation.proxy.ResourceHandlerRequest
import software.amazon.globalaccelerator.accelerator.AcceleratorClientBuilder.client

/**
 * Create handler implementation for accelerator resource.
 */
class CreateHandler : BaseHandler<CallbackContext?>() {
    override fun handleRequest(proxy: AmazonWebServicesClientProxy,
                               request: ResourceHandlerRequest<ResourceModel>,
                               callbackContext: CallbackContext?,
                               logger: Logger): ProgressEvent<ResourceModel, CallbackContext?> {
        logger.debug("Create Accelerator Request: $request")
        val agaClient = client
        val inferredCallbackContext = callbackContext ?: CallbackContext(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
        val model = request.desiredResourceState
        return when (model.acceleratorArn) {
            null -> createAcceleratorStep(model, request, proxy, agaClient, logger)
            else -> HandlerCommons.waitForSynchronizedStep(inferredCallbackContext, model, proxy, agaClient, logger)
        }
    }

    /**
     * Create an accelerator and create the correct progress continuation context
     */
    private fun createAcceleratorStep(model: ResourceModel,
                                      handlerRequest: ResourceHandlerRequest<ResourceModel>,
                                      proxy: AmazonWebServicesClientProxy,
                                      agaClient: AWSGlobalAccelerator,
                                      logger: Logger): ProgressEvent<ResourceModel, CallbackContext?> {
        logger.debug("Creating new accelerator with model: $model")
        val acc = createAccelerator(model, handlerRequest, proxy, agaClient)
        model.apply {
            this.acceleratorArn = acc.acceleratorArn
            this.dnsName = acc.dnsName
            this.dualStackDnsName = acc.dualStackDnsName
            this.ipAddresses = acc.ipSets?.flatMap { it.ipAddresses }
        }
        val callbackContext: CallbackContext? = CallbackContext(HandlerCommons.NUMBER_OF_STATE_POLL_RETRIES)
        return ProgressEvent.defaultInProgressHandler(callbackContext, 0, model)
    }

    private fun createAccelerator(model: ResourceModel,
                                  handlerRequest: ResourceHandlerRequest<ResourceModel>,
                                  proxy: AmazonWebServicesClientProxy,
                                  agaClient: AWSGlobalAccelerator): Accelerator {
        val convertedTags = model.tags?.map { Tag().withKey(it.key).withValue(it.value) }
        val request = CreateAcceleratorRequest()
                .withName(model.name)
                .withIpAddressType(model.ipAddressType)
                .withEnabled(model.enabled)
                .withIpAddresses(model.ipAddresses)
                .withTags(convertedTags)
                .withIdempotencyToken(handlerRequest.clientRequestToken)
        return proxy.injectCredentialsAndInvoke(request, agaClient::createAccelerator).accelerator
    }
}
